src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java

```diff
@@ -0,0 +1,382 @@
+    }
+
+    private ByteComparable encode(ByteBuffer input)
+    {
+        return isLiteral ? version -> append(ByteSource.of(input, version), ByteSource.TERMINATOR)
+                         : version -> TypeUtil.asComparableBytes(input, validator, version);
+    }
+
+    private ByteComparable decode(ByteComparable term)
+    {
+        return isLiteral ? version -> ByteSourceInverse.unescape(ByteSource.peekable(term.asComparableBytes(version)))
+                         : term;
+
+    }
+
+    private ByteSource append(ByteSource src, int lastByte)
+    {
+        return new ByteSource()
+        {
+            boolean done = false;
+
+            @Override
+            public int next()
+            {
+                if (done)
+                    return END_OF_STREAM;
+                int n = src.next();
+                if (n != END_OF_STREAM)
+                    return n;
+
+                done = true;
+                return lastByte;
+            }
+        };
+    }
+
+    private RangeIterator exactMatch(Expression expression)
+    {
+        final ByteComparable prefix = expression.lower == null ? ByteComparable.EMPTY : encode(expression.lower.value.encoded);
+        final PrimaryKeys primaryKeys = data.get(prefix);
+        if (primaryKeys == null)
+        {
+            return RangeIterator.empty();
+        }
+        return new KeyRangeIterator(primaryKeys.keys());
+    }
+
+    public static class Collector
+    {
+        PrimaryKey minimumKey = null;
+        PrimaryKey maximumKey = null;
+        PriorityQueue<PrimaryKey> mergedKeys = new PriorityQueue<>(lastQueueSize.get());
+
+        AbstractBounds<PartitionPosition> keyRange;
+
+        public Collector(AbstractBounds<PartitionPosition> keyRange)
+        {
+            this.keyRange = keyRange;
+        }
+
+        public void processContent(PrimaryKeys keys)
+        {
+            if (keys.isEmpty())
+                return;
+
+            SortedSet<PrimaryKey> primaryKeys = keys.keys();
+
+            // shortcut to avoid generating iterator
+            if (primaryKeys.size() == 1)
+            {
+                PrimaryKey first = primaryKeys.first();
+                if (keyRange.contains(first.partitionKey()))
+                {
+                    mergedKeys.add(first);
+
+                    minimumKey = minimumKey == null ? first : first.compareTo(minimumKey) < 0 ? first : minimumKey;
+                    maximumKey = maximumKey == null ? first : first.compareTo(maximumKey) > 0 ? first : maximumKey;
+                }
+
+                return;
+            }
+
+            // skip entire partition keys if they don't overlap
+            if (!keyRange.right.isMinimum() && primaryKeys.first().partitionKey().compareTo(keyRange.right) > 0
+                || primaryKeys.last().partitionKey().compareTo(keyRange.left) < 0)
+                return;
+
+            for (PrimaryKey key : primaryKeys)
```

Comment:
I've been trying to sniff this out w/ test coverage tools, but are there any tests that get this far in the logic?

Reply:
Correct. I have added a test for this taken from another patch. This adds a `randomQueryTest` that exercises the `add` and `search` paths in a random manner. This also necessitates bringing in a `FilteringInMemoryKeyRangeIterator` for `exactMatch`. The current implementation of `exactMatch` did not check that the keys returned were in the `keyRange` asked for and relied on post-filtering to remove keys outside the range. `FilteringInMemoryKeyRangeIterator` is an extension of `InMemoryKeyRangeIterator` that filters the keys against the key range and only returns the in-range keys. Apart from allowing the `randomQueryTest` to pass this also introduces a small performance improvement.